{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### WIP Methylation Statistical Analysis\n",
    "\n",
    "This code analyzes a multi-tissue DNA methylation dataset from mice, sourced from ['Multi-tissue DNA methylation age predictor in mouse'](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-017-1203-5). It calculates average methylation rate, depth, and identifies tissue-specific and age-related differentiations in the samples. By employing  statistical techniques, this code will provide valuable insights into DNA methylation patterns and their association with aging processes. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "import os\n",
    "import pandas as pd\n",
    "import sys\n",
    "from typing import Tuple, List\n",
    "from tabulate import tabulate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_rows', None)\n",
    "directory = 'data/GSE93957_RAW/'"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load Data & Build Dataframe\n",
    "\n",
    "The build_df function generates a pandas DataFrame from a specified sample name. It initializes an empty DataFrame and defines column names. It iterates through files in the directory, checking for gzip format and matching sample name. Data is read using read_csv, with tab as separator, and stored in a temporary DataFrame. If the main DataFrame is empty, the data is directly assigned. The function returns the compiled DataFrame from matching files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "sample_list = [\"Lung\", \"Heart\", \"Liver\", \"Cortex\"]\n",
    "\n",
    "def build_df(filename) -> pd.DataFrame:\n",
    "    # Create empty DataFrame to be filled\n",
    "    df = pd.DataFrame()\n",
    "    columns_names = ['chromosome', 's_loc', 'e_loc', 'methyl_rate', 's_depth', 'e_depth']\n",
    "    \n",
    "    # # Choose preset file to perform statistical analysis on\n",
    "    # if test:\n",
    "    #     filename = \"GSM2465667_M04NB_1wk_Liver.cov.txt.gz\"\n",
    "    #     columns_names = ['chromosome', 's_loc', 'e_loc', 'methyl_rate', 's_depth', 'e_depth']\n",
    "    #     with gzip.open('data/GSE93957_RAW/GSM2465667_M04NB_1wk_Liver.cov.txt.gz', 'rt') as file: \n",
    "    #         df = pd.read_csv('data/GSE93957_RAW/GSM2465667_M04NB_1wk_Liver.cov.txt.gz', sep=\"\\t\", header=None, names=columns_names, low_memory=False)\n",
    "    #     return df, filename\n",
    "\n",
    "\n",
    "    file_path = os.path.join(directory, filename)\n",
    "    # Open the gzipped file in text mode\n",
    "    with gzip.open(file_path, 'rt') as file:  \n",
    "        print(file_path)\n",
    "        df = pd.read_csv(file_path, sep=\"\\t\", header=None, names=columns_names, low_memory=False)\n",
    "      \n",
    "    return df"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define Functions to parse data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_result(test = False):\n",
    "    column_names = ['id', 'age', 'tissue', 'num_sites', 'ave depth', 'ave methylation', 'ave methylation > 2 depth', 'ave methylation > 5 depth']\n",
    "    result = pd.DataFrame(columns=column_names)\n",
    "    \n",
    "    \n",
    "    if not test:\n",
    "        for filename in os.listdir(directory):\n",
    "            _build_result_helper(filename, result)\n",
    "    else: \n",
    "        filename = \"GSM2465667_M04NB_1wk_Liver.cov.txt.gz\"\n",
    "        _build_result_helper(filename, result)\n",
    "        \n",
    "    return result\n",
    "\n",
    "def _build_result_helper(filename, result):\n",
    "    df = build_df(filename)\n",
    "    length = len(df)\n",
    "    depth = (df['s_depth'] + df['e_depth']).mean()\n",
    "    mean = df['methyl_rate'].mean()\n",
    "    \n",
    "    # Get the average methylation for rows with read_count >= 2 & read_count <= 100\n",
    "    df = df.drop(df[df['e_depth'] + df['s_depth'] < 2].index)\n",
    "    df = df.drop(df[df['e_depth'] + df['s_depth'] > 100].index)\n",
    "    two_depth_mean = df['methyl_rate'].mean()\n",
    "\n",
    "    # Get the average methylation for rows with read_count >= 5 & read_count <= 100\n",
    "    df = df.drop(df[df['e_depth'] + df['s_depth'] < 5].index)\n",
    "    five_depth_mean = df['methyl_rate'].mean()\n",
    "\n",
    "    # Split the filename into id and age_sample\n",
    "    splits = filename.split(\"_\")\n",
    "    id = splits[0]+\"_\"+splits[1]\n",
    "    age = splits[2]\n",
    "    tissue = splits[3].split(\".\")[0]\n",
    "    result.loc[len(result)] = [id, age, tissue, length, depth, mean, two_depth_mean, five_depth_mean]\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load the data and parse the results\n",
    "\n",
    "Parsing through all the data takes about two minutes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "data/GSE93957_RAW/GSM2465633_M00018724_27wk_Liver.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465619_M00018362_41wk_Cortex.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465667_M04NB_1wk_Liver.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465627_M00018381_41wk_Cortex.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465631_M00018724_27wk_Cortex.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465653_M02NB_1wk_Lung.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465636_M00018752_27wk_Heart.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465632_M00018724_27wk_Heart.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465637_M00018752_27wk_Liver.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465625_M00018363_41wk_Liver.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465668_M04NB_1wk_Lung.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465662_M0420527_14wk_Heart.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465665_M04NB_1wk_Cortex.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465642_M00018754_27wk_Lung.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465676_M0520522_14wk_Lung.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465621_M00018362_41wk_Liver.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465675_M0520522_14wk_Liver.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465648_M01NB_1wk_Heart.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465640_M00018754_27wk_Heart.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465646_M0112371_27wk_Lung.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465664_M0420527_14wk_Lung.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465629_M00018381_41wk_Liver.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465656_M03NB_1wk_Lung.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465659_M0420525_14wk_Liver.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465654_M03NB_1wk_Cortex.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465650_M01NB_1wk_Lung.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465663_M0420527_14wk_Liver.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465657_M0420525_14wk_Cortex.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465670_M0512375_27wk_Heart.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465626_M00018363_41wk_Lung.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465623_M00018363_41wk_Cortex.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465620_M00018362_41wk_Heart.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465658_M0420525_14wk_Heart.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465672_M0512375_27wk_Lung.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465644_M0112371_27wk_Heart.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465669_M0512375_27wk_Cortex.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465674_M0520522_14wk_Heart.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465624_M00018363_41wk_Heart.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465634_M00018724_27wk_Lung.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465652_M02NB_1wk_Liver.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465645_M0112371_27wk_Liver.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465630_M00018381_41wk_Lung.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465639_M00018754_27wk_Cortex.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465635_M00018752_27wk_Cortex.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465616_M00018359_41wk_Heart.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465638_M00018752_27wk_Lung.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465673_M0520522_14wk_Cortex.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465651_M02NB_1wk_Cortex.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465641_M00018754_27wk_Liver.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465643_M0112371_27wk_Cortex.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465661_M0420527_14wk_Cortex.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465622_M00018362_41wk_Lung.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465615_M00018359_41wk_Cortex.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465617_M00018359_41wk_Liver.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465660_M0420525_14wk_Lung.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465618_M00018359_41wk_Lung.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465671_M0512375_27wk_Liver.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465628_M00018381_41wk_Heart.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465647_M01NB_1wk_Cortex.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465649_M01NB_1wk_Liver.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465666_M04NB_1wk_Heart.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465655_M03NB_1wk_Heart.cov.txt.gz\n"
     ]
    }
   ],
   "source": [
    "# Perform statistical analysis and build a 'result DataFrame' that will be saved\n",
    "df = build_result()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>age</th>\n",
       "      <th>tissue</th>\n",
       "      <th>num_sites</th>\n",
       "      <th>ave depth</th>\n",
       "      <th>ave methylation</th>\n",
       "      <th>ave methylation &gt; 2 depth</th>\n",
       "      <th>ave methylation &gt; 5 depth</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GSM2465674_M0520522</td>\n",
       "      <td>14wk</td>\n",
       "      <td>Heart</td>\n",
       "      <td>4719569</td>\n",
       "      <td>26.478823</td>\n",
       "      <td>38.266097</td>\n",
       "      <td>34.152285</td>\n",
       "      <td>34.951853</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GSM2465658_M0420525</td>\n",
       "      <td>14wk</td>\n",
       "      <td>Heart</td>\n",
       "      <td>4650887</td>\n",
       "      <td>12.917845</td>\n",
       "      <td>39.010798</td>\n",
       "      <td>34.896987</td>\n",
       "      <td>36.529763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GSM2465673_M0520522</td>\n",
       "      <td>14wk</td>\n",
       "      <td>Cortex</td>\n",
       "      <td>3993331</td>\n",
       "      <td>12.758077</td>\n",
       "      <td>37.903837</td>\n",
       "      <td>36.610926</td>\n",
       "      <td>37.846018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GSM2465676_M0520522</td>\n",
       "      <td>14wk</td>\n",
       "      <td>Lung</td>\n",
       "      <td>3833871</td>\n",
       "      <td>12.380396</td>\n",
       "      <td>36.179557</td>\n",
       "      <td>35.632827</td>\n",
       "      <td>38.103973</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GSM2465657_M0420525</td>\n",
       "      <td>14wk</td>\n",
       "      <td>Cortex</td>\n",
       "      <td>4205076</td>\n",
       "      <td>14.687372</td>\n",
       "      <td>38.739643</td>\n",
       "      <td>36.724608</td>\n",
       "      <td>38.331880</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GSM2465663_M0420527</td>\n",
       "      <td>14wk</td>\n",
       "      <td>Liver</td>\n",
       "      <td>4239585</td>\n",
       "      <td>10.794067</td>\n",
       "      <td>38.554164</td>\n",
       "      <td>35.594224</td>\n",
       "      <td>36.769701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GSM2465660_M0420525</td>\n",
       "      <td>14wk</td>\n",
       "      <td>Lung</td>\n",
       "      <td>4143370</td>\n",
       "      <td>14.824189</td>\n",
       "      <td>36.765770</td>\n",
       "      <td>34.756769</td>\n",
       "      <td>36.027765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GSM2465661_M0420527</td>\n",
       "      <td>14wk</td>\n",
       "      <td>Cortex</td>\n",
       "      <td>4353542</td>\n",
       "      <td>19.095015</td>\n",
       "      <td>38.545853</td>\n",
       "      <td>36.233313</td>\n",
       "      <td>37.388675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GSM2465675_M0520522</td>\n",
       "      <td>14wk</td>\n",
       "      <td>Liver</td>\n",
       "      <td>3984791</td>\n",
       "      <td>9.920883</td>\n",
       "      <td>37.102108</td>\n",
       "      <td>35.713960</td>\n",
       "      <td>37.235556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>GSM2465664_M0420527</td>\n",
       "      <td>14wk</td>\n",
       "      <td>Lung</td>\n",
       "      <td>4017025</td>\n",
       "      <td>14.191690</td>\n",
       "      <td>36.445273</td>\n",
       "      <td>35.236357</td>\n",
       "      <td>37.021787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>GSM2465659_M0420525</td>\n",
       "      <td>14wk</td>\n",
       "      <td>Liver</td>\n",
       "      <td>4251425</td>\n",
       "      <td>13.185862</td>\n",
       "      <td>38.163836</td>\n",
       "      <td>35.632308</td>\n",
       "      <td>37.079572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>GSM2465662_M0420527</td>\n",
       "      <td>14wk</td>\n",
       "      <td>Heart</td>\n",
       "      <td>4012254</td>\n",
       "      <td>11.203590</td>\n",
       "      <td>36.182139</td>\n",
       "      <td>34.516667</td>\n",
       "      <td>35.787269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>GSM2465655_M03NB</td>\n",
       "      <td>1wk</td>\n",
       "      <td>Heart</td>\n",
       "      <td>3864703</td>\n",
       "      <td>14.244776</td>\n",
       "      <td>34.843568</td>\n",
       "      <td>34.219407</td>\n",
       "      <td>36.359592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>GSM2465654_M03NB</td>\n",
       "      <td>1wk</td>\n",
       "      <td>Cortex</td>\n",
       "      <td>4102320</td>\n",
       "      <td>15.945711</td>\n",
       "      <td>36.759149</td>\n",
       "      <td>35.023107</td>\n",
       "      <td>36.283298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>GSM2465652_M02NB</td>\n",
       "      <td>1wk</td>\n",
       "      <td>Liver</td>\n",
       "      <td>4005207</td>\n",
       "      <td>14.615569</td>\n",
       "      <td>34.873195</td>\n",
       "      <td>33.872771</td>\n",
       "      <td>35.290571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>GSM2465666_M04NB</td>\n",
       "      <td>1wk</td>\n",
       "      <td>Heart</td>\n",
       "      <td>4257344</td>\n",
       "      <td>19.341303</td>\n",
       "      <td>34.964655</td>\n",
       "      <td>32.868154</td>\n",
       "      <td>33.876040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>GSM2465656_M03NB</td>\n",
       "      <td>1wk</td>\n",
       "      <td>Lung</td>\n",
       "      <td>4472508</td>\n",
       "      <td>26.167893</td>\n",
       "      <td>36.041643</td>\n",
       "      <td>33.068822</td>\n",
       "      <td>33.736718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>GSM2465650_M01NB</td>\n",
       "      <td>1wk</td>\n",
       "      <td>Lung</td>\n",
       "      <td>3706397</td>\n",
       "      <td>7.522344</td>\n",
       "      <td>34.557359</td>\n",
       "      <td>33.601671</td>\n",
       "      <td>35.058591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>GSM2465648_M01NB</td>\n",
       "      <td>1wk</td>\n",
       "      <td>Heart</td>\n",
       "      <td>4021292</td>\n",
       "      <td>14.609889</td>\n",
       "      <td>35.120765</td>\n",
       "      <td>33.674775</td>\n",
       "      <td>35.392383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>GSM2465651_M02NB</td>\n",
       "      <td>1wk</td>\n",
       "      <td>Cortex</td>\n",
       "      <td>3877028</td>\n",
       "      <td>10.207582</td>\n",
       "      <td>35.922012</td>\n",
       "      <td>34.848987</td>\n",
       "      <td>36.031324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>GSM2465665_M04NB</td>\n",
       "      <td>1wk</td>\n",
       "      <td>Cortex</td>\n",
       "      <td>4353634</td>\n",
       "      <td>14.978642</td>\n",
       "      <td>38.081742</td>\n",
       "      <td>35.307351</td>\n",
       "      <td>36.508318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>GSM2465668_M04NB</td>\n",
       "      <td>1wk</td>\n",
       "      <td>Lung</td>\n",
       "      <td>4873812</td>\n",
       "      <td>23.884442</td>\n",
       "      <td>37.666862</td>\n",
       "      <td>33.448200</td>\n",
       "      <td>33.961155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>GSM2465653_M02NB</td>\n",
       "      <td>1wk</td>\n",
       "      <td>Lung</td>\n",
       "      <td>4085551</td>\n",
       "      <td>15.241029</td>\n",
       "      <td>35.427501</td>\n",
       "      <td>33.872858</td>\n",
       "      <td>35.458821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>GSM2465647_M01NB</td>\n",
       "      <td>1wk</td>\n",
       "      <td>Cortex</td>\n",
       "      <td>3939791</td>\n",
       "      <td>9.395831</td>\n",
       "      <td>36.255070</td>\n",
       "      <td>34.480659</td>\n",
       "      <td>35.639115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>GSM2465667_M04NB</td>\n",
       "      <td>1wk</td>\n",
       "      <td>Liver</td>\n",
       "      <td>4147692</td>\n",
       "      <td>13.342087</td>\n",
       "      <td>36.218352</td>\n",
       "      <td>34.165614</td>\n",
       "      <td>35.222294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>GSM2465649_M01NB</td>\n",
       "      <td>1wk</td>\n",
       "      <td>Liver</td>\n",
       "      <td>3669961</td>\n",
       "      <td>7.935134</td>\n",
       "      <td>34.789121</td>\n",
       "      <td>34.387766</td>\n",
       "      <td>36.722110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>GSM2465635_M00018752</td>\n",
       "      <td>27wk</td>\n",
       "      <td>Cortex</td>\n",
       "      <td>4189271</td>\n",
       "      <td>17.318348</td>\n",
       "      <td>38.355115</td>\n",
       "      <td>36.506813</td>\n",
       "      <td>37.901619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>GSM2465639_M00018754</td>\n",
       "      <td>27wk</td>\n",
       "      <td>Cortex</td>\n",
       "      <td>4182114</td>\n",
       "      <td>15.634243</td>\n",
       "      <td>38.888374</td>\n",
       "      <td>37.268956</td>\n",
       "      <td>39.113288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>GSM2465641_M00018754</td>\n",
       "      <td>27wk</td>\n",
       "      <td>Liver</td>\n",
       "      <td>4160734</td>\n",
       "      <td>11.820936</td>\n",
       "      <td>38.169075</td>\n",
       "      <td>35.667510</td>\n",
       "      <td>37.296568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>GSM2465643_M0112371</td>\n",
       "      <td>27wk</td>\n",
       "      <td>Cortex</td>\n",
       "      <td>3980861</td>\n",
       "      <td>13.733747</td>\n",
       "      <td>37.598453</td>\n",
       "      <td>36.490056</td>\n",
       "      <td>37.380998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>GSM2465671_M0512375</td>\n",
       "      <td>27wk</td>\n",
       "      <td>Liver</td>\n",
       "      <td>3555333</td>\n",
       "      <td>5.994593</td>\n",
       "      <td>36.699378</td>\n",
       "      <td>36.481482</td>\n",
       "      <td>38.411934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>GSM2465645_M0112371</td>\n",
       "      <td>27wk</td>\n",
       "      <td>Liver</td>\n",
       "      <td>3596592</td>\n",
       "      <td>7.141513</td>\n",
       "      <td>37.192085</td>\n",
       "      <td>37.176904</td>\n",
       "      <td>41.087564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>GSM2465634_M00018724</td>\n",
       "      <td>27wk</td>\n",
       "      <td>Lung</td>\n",
       "      <td>4070360</td>\n",
       "      <td>12.648653</td>\n",
       "      <td>37.220572</td>\n",
       "      <td>35.887891</td>\n",
       "      <td>38.637895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>GSM2465638_M00018752</td>\n",
       "      <td>27wk</td>\n",
       "      <td>Lung</td>\n",
       "      <td>4017936</td>\n",
       "      <td>11.916289</td>\n",
       "      <td>37.329438</td>\n",
       "      <td>36.136646</td>\n",
       "      <td>38.724930</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>GSM2465669_M0512375</td>\n",
       "      <td>27wk</td>\n",
       "      <td>Cortex</td>\n",
       "      <td>4136638</td>\n",
       "      <td>15.853547</td>\n",
       "      <td>37.601744</td>\n",
       "      <td>35.818608</td>\n",
       "      <td>35.867736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>GSM2465633_M00018724</td>\n",
       "      <td>27wk</td>\n",
       "      <td>Liver</td>\n",
       "      <td>4748562</td>\n",
       "      <td>11.236037</td>\n",
       "      <td>42.379657</td>\n",
       "      <td>36.123860</td>\n",
       "      <td>36.826557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>GSM2465672_M0512375</td>\n",
       "      <td>27wk</td>\n",
       "      <td>Lung</td>\n",
       "      <td>3911071</td>\n",
       "      <td>11.994986</td>\n",
       "      <td>36.291409</td>\n",
       "      <td>35.309406</td>\n",
       "      <td>37.206601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>GSM2465670_M0512375</td>\n",
       "      <td>27wk</td>\n",
       "      <td>Heart</td>\n",
       "      <td>3971368</td>\n",
       "      <td>16.665973</td>\n",
       "      <td>36.072164</td>\n",
       "      <td>34.944992</td>\n",
       "      <td>36.636785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>GSM2465631_M00018724</td>\n",
       "      <td>27wk</td>\n",
       "      <td>Cortex</td>\n",
       "      <td>4086565</td>\n",
       "      <td>15.372909</td>\n",
       "      <td>38.633424</td>\n",
       "      <td>37.273794</td>\n",
       "      <td>39.312597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>GSM2465646_M0112371</td>\n",
       "      <td>27wk</td>\n",
       "      <td>Lung</td>\n",
       "      <td>3778821</td>\n",
       "      <td>9.500211</td>\n",
       "      <td>36.927381</td>\n",
       "      <td>36.563379</td>\n",
       "      <td>39.890076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>GSM2465636_M00018752</td>\n",
       "      <td>27wk</td>\n",
       "      <td>Heart</td>\n",
       "      <td>4086515</td>\n",
       "      <td>12.402298</td>\n",
       "      <td>35.716553</td>\n",
       "      <td>34.113677</td>\n",
       "      <td>35.165530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>GSM2465640_M00018754</td>\n",
       "      <td>27wk</td>\n",
       "      <td>Heart</td>\n",
       "      <td>3906528</td>\n",
       "      <td>10.429799</td>\n",
       "      <td>35.507349</td>\n",
       "      <td>34.501596</td>\n",
       "      <td>35.713762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>GSM2465642_M00018754</td>\n",
       "      <td>27wk</td>\n",
       "      <td>Lung</td>\n",
       "      <td>3964959</td>\n",
       "      <td>12.814701</td>\n",
       "      <td>36.124068</td>\n",
       "      <td>35.174017</td>\n",
       "      <td>36.823595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>GSM2465644_M0112371</td>\n",
       "      <td>27wk</td>\n",
       "      <td>Heart</td>\n",
       "      <td>4245053</td>\n",
       "      <td>14.198697</td>\n",
       "      <td>36.431059</td>\n",
       "      <td>34.289872</td>\n",
       "      <td>35.496465</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>GSM2465637_M00018752</td>\n",
       "      <td>27wk</td>\n",
       "      <td>Liver</td>\n",
       "      <td>3681091</td>\n",
       "      <td>4.770938</td>\n",
       "      <td>38.885797</td>\n",
       "      <td>36.967356</td>\n",
       "      <td>41.465177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>GSM2465632_M00018724</td>\n",
       "      <td>27wk</td>\n",
       "      <td>Heart</td>\n",
       "      <td>3963909</td>\n",
       "      <td>10.840487</td>\n",
       "      <td>36.595488</td>\n",
       "      <td>35.694637</td>\n",
       "      <td>38.634155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>GSM2465619_M00018362</td>\n",
       "      <td>41wk</td>\n",
       "      <td>Cortex</td>\n",
       "      <td>4121662</td>\n",
       "      <td>14.977096</td>\n",
       "      <td>38.715772</td>\n",
       "      <td>37.048571</td>\n",
       "      <td>38.045658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>GSM2465627_M00018381</td>\n",
       "      <td>41wk</td>\n",
       "      <td>Cortex</td>\n",
       "      <td>4165468</td>\n",
       "      <td>15.123178</td>\n",
       "      <td>37.681559</td>\n",
       "      <td>36.102748</td>\n",
       "      <td>37.034388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>GSM2465628_M00018381</td>\n",
       "      <td>41wk</td>\n",
       "      <td>Heart</td>\n",
       "      <td>3917926</td>\n",
       "      <td>12.731100</td>\n",
       "      <td>35.657432</td>\n",
       "      <td>34.545901</td>\n",
       "      <td>35.690143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>GSM2465618_M00018359</td>\n",
       "      <td>41wk</td>\n",
       "      <td>Lung</td>\n",
       "      <td>3994127</td>\n",
       "      <td>13.631188</td>\n",
       "      <td>37.365129</td>\n",
       "      <td>36.303361</td>\n",
       "      <td>38.738446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>GSM2465617_M00018359</td>\n",
       "      <td>41wk</td>\n",
       "      <td>Liver</td>\n",
       "      <td>4251469</td>\n",
       "      <td>8.114054</td>\n",
       "      <td>40.625131</td>\n",
       "      <td>36.382638</td>\n",
       "      <td>38.775739</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>GSM2465615_M00018359</td>\n",
       "      <td>41wk</td>\n",
       "      <td>Cortex</td>\n",
       "      <td>4085575</td>\n",
       "      <td>14.700237</td>\n",
       "      <td>38.033469</td>\n",
       "      <td>36.744396</td>\n",
       "      <td>38.389142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>GSM2465625_M00018363</td>\n",
       "      <td>41wk</td>\n",
       "      <td>Liver</td>\n",
       "      <td>4546773</td>\n",
       "      <td>16.950072</td>\n",
       "      <td>39.216508</td>\n",
       "      <td>35.290179</td>\n",
       "      <td>35.738288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>GSM2465621_M00018362</td>\n",
       "      <td>41wk</td>\n",
       "      <td>Liver</td>\n",
       "      <td>3823093</td>\n",
       "      <td>5.591820</td>\n",
       "      <td>38.610837</td>\n",
       "      <td>35.921888</td>\n",
       "      <td>38.461123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>GSM2465616_M00018359</td>\n",
       "      <td>41wk</td>\n",
       "      <td>Heart</td>\n",
       "      <td>4277548</td>\n",
       "      <td>14.498222</td>\n",
       "      <td>36.996101</td>\n",
       "      <td>34.584712</td>\n",
       "      <td>35.891512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>GSM2465629_M00018381</td>\n",
       "      <td>41wk</td>\n",
       "      <td>Liver</td>\n",
       "      <td>3941176</td>\n",
       "      <td>6.002234</td>\n",
       "      <td>39.283386</td>\n",
       "      <td>37.027295</td>\n",
       "      <td>40.523243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>GSM2465630_M00018381</td>\n",
       "      <td>41wk</td>\n",
       "      <td>Lung</td>\n",
       "      <td>4030714</td>\n",
       "      <td>13.589659</td>\n",
       "      <td>36.746636</td>\n",
       "      <td>35.558433</td>\n",
       "      <td>37.582641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>GSM2465626_M00018363</td>\n",
       "      <td>41wk</td>\n",
       "      <td>Lung</td>\n",
       "      <td>4210256</td>\n",
       "      <td>25.553273</td>\n",
       "      <td>36.127743</td>\n",
       "      <td>34.469923</td>\n",
       "      <td>35.334318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>GSM2465620_M00018362</td>\n",
       "      <td>41wk</td>\n",
       "      <td>Heart</td>\n",
       "      <td>4572977</td>\n",
       "      <td>13.387098</td>\n",
       "      <td>38.420207</td>\n",
       "      <td>34.901906</td>\n",
       "      <td>36.381228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>GSM2465624_M00018363</td>\n",
       "      <td>41wk</td>\n",
       "      <td>Heart</td>\n",
       "      <td>4348186</td>\n",
       "      <td>17.300112</td>\n",
       "      <td>36.629723</td>\n",
       "      <td>34.209188</td>\n",
       "      <td>35.232790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>GSM2465622_M00018362</td>\n",
       "      <td>41wk</td>\n",
       "      <td>Lung</td>\n",
       "      <td>3980174</td>\n",
       "      <td>12.756365</td>\n",
       "      <td>36.455039</td>\n",
       "      <td>35.213992</td>\n",
       "      <td>36.593348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>GSM2465623_M00018363</td>\n",
       "      <td>41wk</td>\n",
       "      <td>Cortex</td>\n",
       "      <td>4078447</td>\n",
       "      <td>14.662973</td>\n",
       "      <td>37.689944</td>\n",
       "      <td>36.139043</td>\n",
       "      <td>36.960516</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      id   age  tissue  num_sites  ave depth  ave methylation  \\\n",
       "0    GSM2465674_M0520522  14wk   Heart    4719569  26.478823        38.266097   \n",
       "1    GSM2465658_M0420525  14wk   Heart    4650887  12.917845        39.010798   \n",
       "2    GSM2465673_M0520522  14wk  Cortex    3993331  12.758077        37.903837   \n",
       "3    GSM2465676_M0520522  14wk    Lung    3833871  12.380396        36.179557   \n",
       "4    GSM2465657_M0420525  14wk  Cortex    4205076  14.687372        38.739643   \n",
       "5    GSM2465663_M0420527  14wk   Liver    4239585  10.794067        38.554164   \n",
       "6    GSM2465660_M0420525  14wk    Lung    4143370  14.824189        36.765770   \n",
       "7    GSM2465661_M0420527  14wk  Cortex    4353542  19.095015        38.545853   \n",
       "8    GSM2465675_M0520522  14wk   Liver    3984791   9.920883        37.102108   \n",
       "9    GSM2465664_M0420527  14wk    Lung    4017025  14.191690        36.445273   \n",
       "10   GSM2465659_M0420525  14wk   Liver    4251425  13.185862        38.163836   \n",
       "11   GSM2465662_M0420527  14wk   Heart    4012254  11.203590        36.182139   \n",
       "12      GSM2465655_M03NB   1wk   Heart    3864703  14.244776        34.843568   \n",
       "13      GSM2465654_M03NB   1wk  Cortex    4102320  15.945711        36.759149   \n",
       "14      GSM2465652_M02NB   1wk   Liver    4005207  14.615569        34.873195   \n",
       "15      GSM2465666_M04NB   1wk   Heart    4257344  19.341303        34.964655   \n",
       "16      GSM2465656_M03NB   1wk    Lung    4472508  26.167893        36.041643   \n",
       "17      GSM2465650_M01NB   1wk    Lung    3706397   7.522344        34.557359   \n",
       "18      GSM2465648_M01NB   1wk   Heart    4021292  14.609889        35.120765   \n",
       "19      GSM2465651_M02NB   1wk  Cortex    3877028  10.207582        35.922012   \n",
       "20      GSM2465665_M04NB   1wk  Cortex    4353634  14.978642        38.081742   \n",
       "21      GSM2465668_M04NB   1wk    Lung    4873812  23.884442        37.666862   \n",
       "22      GSM2465653_M02NB   1wk    Lung    4085551  15.241029        35.427501   \n",
       "23      GSM2465647_M01NB   1wk  Cortex    3939791   9.395831        36.255070   \n",
       "24      GSM2465667_M04NB   1wk   Liver    4147692  13.342087        36.218352   \n",
       "25      GSM2465649_M01NB   1wk   Liver    3669961   7.935134        34.789121   \n",
       "26  GSM2465635_M00018752  27wk  Cortex    4189271  17.318348        38.355115   \n",
       "27  GSM2465639_M00018754  27wk  Cortex    4182114  15.634243        38.888374   \n",
       "28  GSM2465641_M00018754  27wk   Liver    4160734  11.820936        38.169075   \n",
       "29   GSM2465643_M0112371  27wk  Cortex    3980861  13.733747        37.598453   \n",
       "30   GSM2465671_M0512375  27wk   Liver    3555333   5.994593        36.699378   \n",
       "31   GSM2465645_M0112371  27wk   Liver    3596592   7.141513        37.192085   \n",
       "32  GSM2465634_M00018724  27wk    Lung    4070360  12.648653        37.220572   \n",
       "33  GSM2465638_M00018752  27wk    Lung    4017936  11.916289        37.329438   \n",
       "34   GSM2465669_M0512375  27wk  Cortex    4136638  15.853547        37.601744   \n",
       "35  GSM2465633_M00018724  27wk   Liver    4748562  11.236037        42.379657   \n",
       "36   GSM2465672_M0512375  27wk    Lung    3911071  11.994986        36.291409   \n",
       "37   GSM2465670_M0512375  27wk   Heart    3971368  16.665973        36.072164   \n",
       "38  GSM2465631_M00018724  27wk  Cortex    4086565  15.372909        38.633424   \n",
       "39   GSM2465646_M0112371  27wk    Lung    3778821   9.500211        36.927381   \n",
       "40  GSM2465636_M00018752  27wk   Heart    4086515  12.402298        35.716553   \n",
       "41  GSM2465640_M00018754  27wk   Heart    3906528  10.429799        35.507349   \n",
       "42  GSM2465642_M00018754  27wk    Lung    3964959  12.814701        36.124068   \n",
       "43   GSM2465644_M0112371  27wk   Heart    4245053  14.198697        36.431059   \n",
       "44  GSM2465637_M00018752  27wk   Liver    3681091   4.770938        38.885797   \n",
       "45  GSM2465632_M00018724  27wk   Heart    3963909  10.840487        36.595488   \n",
       "46  GSM2465619_M00018362  41wk  Cortex    4121662  14.977096        38.715772   \n",
       "47  GSM2465627_M00018381  41wk  Cortex    4165468  15.123178        37.681559   \n",
       "48  GSM2465628_M00018381  41wk   Heart    3917926  12.731100        35.657432   \n",
       "49  GSM2465618_M00018359  41wk    Lung    3994127  13.631188        37.365129   \n",
       "50  GSM2465617_M00018359  41wk   Liver    4251469   8.114054        40.625131   \n",
       "51  GSM2465615_M00018359  41wk  Cortex    4085575  14.700237        38.033469   \n",
       "52  GSM2465625_M00018363  41wk   Liver    4546773  16.950072        39.216508   \n",
       "53  GSM2465621_M00018362  41wk   Liver    3823093   5.591820        38.610837   \n",
       "54  GSM2465616_M00018359  41wk   Heart    4277548  14.498222        36.996101   \n",
       "55  GSM2465629_M00018381  41wk   Liver    3941176   6.002234        39.283386   \n",
       "56  GSM2465630_M00018381  41wk    Lung    4030714  13.589659        36.746636   \n",
       "57  GSM2465626_M00018363  41wk    Lung    4210256  25.553273        36.127743   \n",
       "58  GSM2465620_M00018362  41wk   Heart    4572977  13.387098        38.420207   \n",
       "59  GSM2465624_M00018363  41wk   Heart    4348186  17.300112        36.629723   \n",
       "60  GSM2465622_M00018362  41wk    Lung    3980174  12.756365        36.455039   \n",
       "61  GSM2465623_M00018363  41wk  Cortex    4078447  14.662973        37.689944   \n",
       "\n",
       "    ave methylation > 2 depth  ave methylation > 5 depth  \n",
       "0                   34.152285                  34.951853  \n",
       "1                   34.896987                  36.529763  \n",
       "2                   36.610926                  37.846018  \n",
       "3                   35.632827                  38.103973  \n",
       "4                   36.724608                  38.331880  \n",
       "5                   35.594224                  36.769701  \n",
       "6                   34.756769                  36.027765  \n",
       "7                   36.233313                  37.388675  \n",
       "8                   35.713960                  37.235556  \n",
       "9                   35.236357                  37.021787  \n",
       "10                  35.632308                  37.079572  \n",
       "11                  34.516667                  35.787269  \n",
       "12                  34.219407                  36.359592  \n",
       "13                  35.023107                  36.283298  \n",
       "14                  33.872771                  35.290571  \n",
       "15                  32.868154                  33.876040  \n",
       "16                  33.068822                  33.736718  \n",
       "17                  33.601671                  35.058591  \n",
       "18                  33.674775                  35.392383  \n",
       "19                  34.848987                  36.031324  \n",
       "20                  35.307351                  36.508318  \n",
       "21                  33.448200                  33.961155  \n",
       "22                  33.872858                  35.458821  \n",
       "23                  34.480659                  35.639115  \n",
       "24                  34.165614                  35.222294  \n",
       "25                  34.387766                  36.722110  \n",
       "26                  36.506813                  37.901619  \n",
       "27                  37.268956                  39.113288  \n",
       "28                  35.667510                  37.296568  \n",
       "29                  36.490056                  37.380998  \n",
       "30                  36.481482                  38.411934  \n",
       "31                  37.176904                  41.087564  \n",
       "32                  35.887891                  38.637895  \n",
       "33                  36.136646                  38.724930  \n",
       "34                  35.818608                  35.867736  \n",
       "35                  36.123860                  36.826557  \n",
       "36                  35.309406                  37.206601  \n",
       "37                  34.944992                  36.636785  \n",
       "38                  37.273794                  39.312597  \n",
       "39                  36.563379                  39.890076  \n",
       "40                  34.113677                  35.165530  \n",
       "41                  34.501596                  35.713762  \n",
       "42                  35.174017                  36.823595  \n",
       "43                  34.289872                  35.496465  \n",
       "44                  36.967356                  41.465177  \n",
       "45                  35.694637                  38.634155  \n",
       "46                  37.048571                  38.045658  \n",
       "47                  36.102748                  37.034388  \n",
       "48                  34.545901                  35.690143  \n",
       "49                  36.303361                  38.738446  \n",
       "50                  36.382638                  38.775739  \n",
       "51                  36.744396                  38.389142  \n",
       "52                  35.290179                  35.738288  \n",
       "53                  35.921888                  38.461123  \n",
       "54                  34.584712                  35.891512  \n",
       "55                  37.027295                  40.523243  \n",
       "56                  35.558433                  37.582641  \n",
       "57                  34.469923                  35.334318  \n",
       "58                  34.901906                  36.381228  \n",
       "59                  34.209188                  35.232790  \n",
       "60                  35.213992                  36.593348  \n",
       "61                  36.139043                  36.960516  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.sort_values(\"age\", ignore_index=True, inplace=True)\n",
    "# Display Results Table in ASCII format:\n",
    "#pdtabulate=lambda df:tabulate(df,headers='keys',tablefmt='psql')\n",
    "#print(pdtabulate(df))\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_4572/2606758392.py\u001b[0m in \u001b[0;36m?\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'age'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;34m'41wk'\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'tissue'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;34m'Cortex'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/Programming/Bioinformatics/Projects/Methylation_1/Bioinformatic_Lab-1/vmeth/lib/python3.10/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1464\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mfinal\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1465\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__nonzero__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0mNoReturn\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1466\u001b[0;31m         raise ValueError(\n\u001b[0m\u001b[1;32m   1467\u001b[0m             \u001b[0;34mf\"The truth value of a {type(self).__name__} is ambiguous. \"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1468\u001b[0m             \u001b[0;34m\"Use a.empty, a.bool(), a.item(), a.any() or a.all().\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1469\u001b[0m         )\n",
      "\u001b[0;31mValueError\u001b[0m: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all()."
     ]
    }
   ],
   "source": [
    "df[(df['age']=='41wk') & (df['tissue']=='Cortex')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save results\n",
    "df.to_csv('result.csv', index=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Methylation rates by age\n",
    "Lets calculate the average methylation rate for each age group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>age</th>\n",
       "      <th>num_sites</th>\n",
       "      <th>ave depth</th>\n",
       "      <th>ave methylation</th>\n",
       "      <th>ave methylation &gt; 2 depth</th>\n",
       "      <th>ave methylation &gt; 5 depth</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>14wk</td>\n",
       "      <td>50404726</td>\n",
       "      <td>14.369817</td>\n",
       "      <td>37.654923</td>\n",
       "      <td>35.475103</td>\n",
       "      <td>36.922818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1wk</td>\n",
       "      <td>57377240</td>\n",
       "      <td>14.816588</td>\n",
       "      <td>35.822928</td>\n",
       "      <td>34.060010</td>\n",
       "      <td>35.395738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>27wk</td>\n",
       "      <td>80234281</td>\n",
       "      <td>12.114445</td>\n",
       "      <td>37.430929</td>\n",
       "      <td>35.919573</td>\n",
       "      <td>37.879692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>41wk</td>\n",
       "      <td>66345571</td>\n",
       "      <td>13.723043</td>\n",
       "      <td>37.765913</td>\n",
       "      <td>35.652761</td>\n",
       "      <td>37.210783</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    age  num_sites  ave depth  ave methylation  ave methylation > 2 depth  \\\n",
       "0  14wk   50404726  14.369817        37.654923                  35.475103   \n",
       "1   1wk   57377240  14.816588        35.822928                  34.060010   \n",
       "2  27wk   80234281  12.114445        37.430929                  35.919573   \n",
       "3  41wk   66345571  13.723043        37.765913                  35.652761   \n",
       "\n",
       "   ave methylation > 5 depth  \n",
       "0                  36.922818  \n",
       "1                  35.395738  \n",
       "2                  37.879692  \n",
       "3                  37.210783  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "column_names = ['age', 'num_sites', 'ave depth', 'ave methylation', 'ave methylation > 2 depth', 'ave methylation > 5 depth']\n",
    "age_df = pd.DataFrame(columns=column_names)\n",
    "\n",
    "age_list = df['age'].unique()\n",
    "for age in age_list:\n",
    "    num_sites = df[df['age'] == age]['num_sites'].sum()\n",
    "    ave_depth = df[df['age'] == age]['ave depth'].mean()\n",
    "    ave_meth = df[df['age'] == age]['ave methylation'].mean()\n",
    "    ave_meth_2 = df[df['age'] == age]['ave methylation > 2 depth'].mean()\n",
    "    ave_meth_5 = df[df['age'] == age]['ave methylation > 5 depth'].mean()\n",
    "    age_df.loc[len(age_df.index)] = [age, num_sites, ave_depth, ave_meth, ave_meth_2, ave_meth_5]\n",
    "\n",
    "age_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "37.168673395674816\n",
      "35.27686151943653\n",
      "36.85225744294009\n"
     ]
    }
   ],
   "source": [
    "print(age_df['ave methylation'].mean())\n",
    "print(age_df['ave methylation > 2 depth'].mean())\n",
    "print(age_df['ave methylation > 5 depth'].mean())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "vmeth",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
