{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "import os\n",
    "import glob\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import sys\n",
    "from typing import Tuple, List\n",
    "from tabulate import tabulate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_rows', None)\n",
    "directory = 'data/GSE93957_RAW/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "sample_list = [\"Lung\", \"Heart\", \"Liver\", \"Cortex\"]\n",
    "\n",
    "def build_df(filename) -> pd.DataFrame:\n",
    "    # Create empty DataFrame to be filled\n",
    "    df = pd.DataFrame()\n",
    "    columns_names = ['chromosome', 's_loc', 'e_loc', 'methyl_rate', 's_depth', 'e_depth']\n",
    "\n",
    "    file_path = os.path.join(directory, filename)\n",
    "    # Open the gzipped file in text mode\n",
    "    with gzip.open(file_path, 'rt') as file:  \n",
    "        print(file_path)\n",
    "        df = pd.read_csv(file_path, sep=\"\\t\", header=None, names=columns_names, low_memory=False)\n",
    "\n",
    "        # Add Column for CG-Sites\n",
    "        loc_list = df['s_loc'].tolist()\n",
    "        cg_list = []\n",
    "        for idx in range(len(loc_list)):\n",
    "            # check if index isn't the last value, and check if next methylated read is sequential to current index\n",
    "            if idx < len(loc_list)-1 and (loc_list[idx] == loc_list[idx+1]-1):\n",
    "                cg_list.append(True)\n",
    "            # check if index isn't the first value, and check if the previous methylated read is sequential to current index\n",
    "            elif idx > 0 and (loc_list[idx] == loc_list[idx-1]+1):\n",
    "                cg_list.append(True)\n",
    "            # Append false otherwise\n",
    "            else:\n",
    "                cg_list.append(False)\n",
    "        df['CG site'] = cg_list\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "data/GSE93957_RAW/GSM2465618_M00018359_41wk_Lung.cov.txt.gz\n"
     ]
    }
   ],
   "source": [
    "df = build_df(\"GSM2465618_M00018359_41wk_Lung.cov.txt.gz\")\n",
    "unique_chromosomes = df['chromosome'].unique()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_chromo_sort = df.sort_values(by=['chromosome', 's_loc']).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>s_loc</th>\n",
       "      <th>e_loc</th>\n",
       "      <th>methyl_rate</th>\n",
       "      <th>s_depth</th>\n",
       "      <th>e_depth</th>\n",
       "      <th>CG site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>3007431</td>\n",
       "      <td>3007431</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>3014611</td>\n",
       "      <td>3014611</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>3014612</td>\n",
       "      <td>3014612</td>\n",
       "      <td>7.692308</td>\n",
       "      <td>2</td>\n",
       "      <td>24</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>3014928</td>\n",
       "      <td>3014928</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>3014929</td>\n",
       "      <td>3014929</td>\n",
       "      <td>76.000000</td>\n",
       "      <td>19</td>\n",
       "      <td>6</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>3014974</td>\n",
       "      <td>3014974</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>3014975</td>\n",
       "      <td>3014975</td>\n",
       "      <td>92.592593</td>\n",
       "      <td>25</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>3020794</td>\n",
       "      <td>3020794</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>3020814</td>\n",
       "      <td>3020814</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>3020842</td>\n",
       "      <td>3020842</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chromosome    s_loc    e_loc  methyl_rate  s_depth  e_depth  CG site\n",
       "0          1  3007431  3007431     0.000000        0        1    False\n",
       "1          1  3014611  3014611   100.000000        1        0     True\n",
       "2          1  3014612  3014612     7.692308        2       24     True\n",
       "3          1  3014928  3014928   100.000000        1        0     True\n",
       "4          1  3014929  3014929    76.000000       19        6     True\n",
       "5          1  3014974  3014974     0.000000        0        1     True\n",
       "6          1  3014975  3014975    92.592593       25        2     True\n",
       "7          1  3020794  3020794   100.000000        1        0    False\n",
       "8          1  3020814  3020814   100.000000        1        0    False\n",
       "9          1  3020842  3020842     0.000000        0        1    False"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_chromo_sort.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "data/GSE93957_RAW/GSM2465622_M00018362_41wk_Lung.cov.txt.gz\n"
     ]
    }
   ],
   "source": [
    "df2 = build_df(\"GSM2465622_M00018362_41wk_Lung.cov.txt.gz\")\n",
    "df_chromo_sort2 = df2.sort_values(by=['chromosome', 's_loc']).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>s_loc</th>\n",
       "      <th>e_loc</th>\n",
       "      <th>methyl_rate</th>\n",
       "      <th>s_depth</th>\n",
       "      <th>e_depth</th>\n",
       "      <th>CG site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>3007431</td>\n",
       "      <td>3007431</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>3007533</td>\n",
       "      <td>3007533</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>3014611</td>\n",
       "      <td>3014611</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>3014612</td>\n",
       "      <td>3014612</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>3014928</td>\n",
       "      <td>3014928</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>3014929</td>\n",
       "      <td>3014929</td>\n",
       "      <td>90.000000</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>3014974</td>\n",
       "      <td>3014974</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>3014975</td>\n",
       "      <td>3014975</td>\n",
       "      <td>90.909091</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>3016622</td>\n",
       "      <td>3016622</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>3020877</td>\n",
       "      <td>3020877</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chromosome    s_loc    e_loc  methyl_rate  s_depth  e_depth  CG site\n",
       "0          1  3007431  3007431     0.000000        0        1    False\n",
       "1          1  3007533  3007533   100.000000        1        0    False\n",
       "2          1  3014611  3014611   100.000000        2        0     True\n",
       "3          1  3014612  3014612     0.000000        0       11     True\n",
       "4          1  3014928  3014928     0.000000        0        1     True\n",
       "5          1  3014929  3014929    90.000000        9        1     True\n",
       "6          1  3014974  3014974     0.000000        0        2     True\n",
       "7          1  3014975  3014975    90.909091       10        1     True\n",
       "8          1  3016622  3016622     0.000000        0        1    False\n",
       "9          1  3020877  3020877   100.000000        7        0     True"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_chromo_sort2.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>s_loc</th>\n",
       "      <th>e_loc_df1</th>\n",
       "      <th>methyl_rate_df1</th>\n",
       "      <th>s_depth_df1</th>\n",
       "      <th>e_depth_df1</th>\n",
       "      <th>CG site_df1</th>\n",
       "      <th>e_loc_df2</th>\n",
       "      <th>methyl_rate_df2</th>\n",
       "      <th>s_depth_df2</th>\n",
       "      <th>e_depth_df2</th>\n",
       "      <th>CG site_df2</th>\n",
       "      <th>avg_methyl_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>3007431</td>\n",
       "      <td>3007431.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>3007431.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>3014611</td>\n",
       "      <td>3014611.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>3014611.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>3014612</td>\n",
       "      <td>3014612.0</td>\n",
       "      <td>7.692308</td>\n",
       "      <td>2.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>True</td>\n",
       "      <td>3014612.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>True</td>\n",
       "      <td>3.846154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>3014928</td>\n",
       "      <td>3014928.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>3014928.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>50.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>3014929</td>\n",
       "      <td>3014929.0</td>\n",
       "      <td>76.000000</td>\n",
       "      <td>19.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>True</td>\n",
       "      <td>3014929.0</td>\n",
       "      <td>90.000000</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>83.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>3014974</td>\n",
       "      <td>3014974.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>3014974.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>3014975</td>\n",
       "      <td>3014975.0</td>\n",
       "      <td>92.592593</td>\n",
       "      <td>25.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>3014975.0</td>\n",
       "      <td>90.909091</td>\n",
       "      <td>10.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>91.750842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>3020794</td>\n",
       "      <td>3020794.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>3020814</td>\n",
       "      <td>3020814.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>3020842</td>\n",
       "      <td>3020842.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1</td>\n",
       "      <td>3020877</td>\n",
       "      <td>3020877.0</td>\n",
       "      <td>80.000000</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>3020877.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>90.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1</td>\n",
       "      <td>3020878</td>\n",
       "      <td>3020878.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>True</td>\n",
       "      <td>3020878.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>True</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1</td>\n",
       "      <td>3020891</td>\n",
       "      <td>3020891.0</td>\n",
       "      <td>40.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>True</td>\n",
       "      <td>3020891.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>70.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1</td>\n",
       "      <td>3020892</td>\n",
       "      <td>3020892.0</td>\n",
       "      <td>75.000000</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>3020892.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>87.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1</td>\n",
       "      <td>3020945</td>\n",
       "      <td>3020945.0</td>\n",
       "      <td>14.285714</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>True</td>\n",
       "      <td>3020945.0</td>\n",
       "      <td>16.666667</td>\n",
       "      <td>2.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>True</td>\n",
       "      <td>15.476190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1</td>\n",
       "      <td>3020946</td>\n",
       "      <td>3020946.0</td>\n",
       "      <td>58.333333</td>\n",
       "      <td>7.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>True</td>\n",
       "      <td>3020946.0</td>\n",
       "      <td>66.666667</td>\n",
       "      <td>10.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>True</td>\n",
       "      <td>62.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1</td>\n",
       "      <td>3020971</td>\n",
       "      <td>3020971.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>3020971.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1</td>\n",
       "      <td>3020972</td>\n",
       "      <td>3020972.0</td>\n",
       "      <td>75.000000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>3020972.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>87.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1</td>\n",
       "      <td>3020987</td>\n",
       "      <td>3020987.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>3020987.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1</td>\n",
       "      <td>3020988</td>\n",
       "      <td>3020988.0</td>\n",
       "      <td>75.000000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>3020988.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>87.500000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chromosome    s_loc  e_loc_df1  methyl_rate_df1  s_depth_df1  e_depth_df1  \\\n",
       "0           1  3007431  3007431.0         0.000000          0.0          1.0   \n",
       "1           1  3014611  3014611.0       100.000000          1.0          0.0   \n",
       "2           1  3014612  3014612.0         7.692308          2.0         24.0   \n",
       "3           1  3014928  3014928.0       100.000000          1.0          0.0   \n",
       "4           1  3014929  3014929.0        76.000000         19.0          6.0   \n",
       "5           1  3014974  3014974.0         0.000000          0.0          1.0   \n",
       "6           1  3014975  3014975.0        92.592593         25.0          2.0   \n",
       "7           1  3020794  3020794.0       100.000000          1.0          0.0   \n",
       "8           1  3020814  3020814.0       100.000000          1.0          0.0   \n",
       "9           1  3020842  3020842.0         0.000000          0.0          1.0   \n",
       "10          1  3020877  3020877.0        80.000000          4.0          1.0   \n",
       "11          1  3020878  3020878.0         0.000000          0.0          8.0   \n",
       "12          1  3020891  3020891.0        40.000000          2.0          3.0   \n",
       "13          1  3020892  3020892.0        75.000000          6.0          2.0   \n",
       "14          1  3020945  3020945.0        14.285714          1.0          6.0   \n",
       "15          1  3020946  3020946.0        58.333333          7.0          5.0   \n",
       "16          1  3020971  3020971.0       100.000000          2.0          0.0   \n",
       "17          1  3020972  3020972.0        75.000000          3.0          1.0   \n",
       "18          1  3020987  3020987.0       100.000000          2.0          0.0   \n",
       "19          1  3020988  3020988.0        75.000000          3.0          1.0   \n",
       "\n",
       "   CG site_df1  e_loc_df2  methyl_rate_df2  s_depth_df2  e_depth_df2  \\\n",
       "0        False  3007431.0         0.000000          0.0          1.0   \n",
       "1         True  3014611.0       100.000000          2.0          0.0   \n",
       "2         True  3014612.0         0.000000          0.0         11.0   \n",
       "3         True  3014928.0         0.000000          0.0          1.0   \n",
       "4         True  3014929.0        90.000000          9.0          1.0   \n",
       "5         True  3014974.0         0.000000          0.0          2.0   \n",
       "6         True  3014975.0        90.909091         10.0          1.0   \n",
       "7        False        NaN              NaN          NaN          NaN   \n",
       "8        False        NaN              NaN          NaN          NaN   \n",
       "9        False        NaN              NaN          NaN          NaN   \n",
       "10        True  3020877.0       100.000000          7.0          0.0   \n",
       "11        True  3020878.0         0.000000          0.0         11.0   \n",
       "12        True  3020891.0       100.000000          9.0          0.0   \n",
       "13        True  3020892.0       100.000000         11.0          0.0   \n",
       "14        True  3020945.0        16.666667          2.0         10.0   \n",
       "15        True  3020946.0        66.666667         10.0          5.0   \n",
       "16        True  3020971.0       100.000000          3.0          0.0   \n",
       "17        True  3020972.0       100.000000          4.0          0.0   \n",
       "18        True  3020987.0       100.000000          3.0          0.0   \n",
       "19        True  3020988.0       100.000000          4.0          0.0   \n",
       "\n",
       "   CG site_df2  avg_methyl_rate  \n",
       "0        False         0.000000  \n",
       "1         True       100.000000  \n",
       "2         True         3.846154  \n",
       "3         True        50.000000  \n",
       "4         True        83.000000  \n",
       "5         True         0.000000  \n",
       "6         True        91.750842  \n",
       "7          NaN              NaN  \n",
       "8          NaN              NaN  \n",
       "9          NaN              NaN  \n",
       "10        True        90.000000  \n",
       "11        True         0.000000  \n",
       "12        True        70.000000  \n",
       "13        True        87.500000  \n",
       "14        True        15.476190  \n",
       "15        True        62.500000  \n",
       "16        True       100.000000  \n",
       "17        True        87.500000  \n",
       "18        True       100.000000  \n",
       "19        True        87.500000  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge the two dataframes on 'chromosome' and 's_loc'\n",
    "merged_df = pd.merge(df_chromo_sort, df_chromo_sort2, on=['chromosome', 's_loc'], how=\"outer\", suffixes=('_df1', '_df2'), sort=True)\n",
    "\n",
    "# Calculate the average of methyl_rate from both dataframes\n",
    "merged_df['avg_methyl_rate'] = (merged_df['methyl_rate_df1'] + merged_df['methyl_rate_df2']) / 2\n",
    "\n",
    "# Drop the individual methyl_rate columns if not needed\n",
    "# merged_df.drop(['methyl_rate_df1', 'methyl_rate_df2'], axis=1, inplace=True)\n",
    "\n",
    "merged_df.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_df['avg_methyl_rate'] = merged_df['avg_methyl_rate'].fillna(merged_df['methyl_rate_df1'].combine_first(merged_df['methyl_rate_df2']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>s_loc</th>\n",
       "      <th>e_loc_df1</th>\n",
       "      <th>methyl_rate_df1</th>\n",
       "      <th>s_depth_df1</th>\n",
       "      <th>e_depth_df1</th>\n",
       "      <th>CG site_df1</th>\n",
       "      <th>e_loc_df2</th>\n",
       "      <th>methyl_rate_df2</th>\n",
       "      <th>s_depth_df2</th>\n",
       "      <th>e_depth_df2</th>\n",
       "      <th>CG site_df2</th>\n",
       "      <th>avg_methyl_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>3007431</td>\n",
       "      <td>3007431.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>3007431.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>3014611</td>\n",
       "      <td>3014611.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>3014611.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>3014612</td>\n",
       "      <td>3014612.0</td>\n",
       "      <td>7.692308</td>\n",
       "      <td>2.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>True</td>\n",
       "      <td>3014612.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>True</td>\n",
       "      <td>3.846154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>3014928</td>\n",
       "      <td>3014928.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>3014928.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>50.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>3014929</td>\n",
       "      <td>3014929.0</td>\n",
       "      <td>76.000000</td>\n",
       "      <td>19.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>True</td>\n",
       "      <td>3014929.0</td>\n",
       "      <td>90.000000</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>83.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>3014974</td>\n",
       "      <td>3014974.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>3014974.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>3014975</td>\n",
       "      <td>3014975.0</td>\n",
       "      <td>92.592593</td>\n",
       "      <td>25.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>3014975.0</td>\n",
       "      <td>90.909091</td>\n",
       "      <td>10.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>91.750842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>3020794</td>\n",
       "      <td>3020794.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>3020814</td>\n",
       "      <td>3020814.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>3020842</td>\n",
       "      <td>3020842.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1</td>\n",
       "      <td>3020877</td>\n",
       "      <td>3020877.0</td>\n",
       "      <td>80.000000</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>3020877.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>90.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1</td>\n",
       "      <td>3020878</td>\n",
       "      <td>3020878.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>True</td>\n",
       "      <td>3020878.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>True</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1</td>\n",
       "      <td>3020891</td>\n",
       "      <td>3020891.0</td>\n",
       "      <td>40.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>True</td>\n",
       "      <td>3020891.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>70.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1</td>\n",
       "      <td>3020892</td>\n",
       "      <td>3020892.0</td>\n",
       "      <td>75.000000</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>True</td>\n",
       "      <td>3020892.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>87.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1</td>\n",
       "      <td>3020945</td>\n",
       "      <td>3020945.0</td>\n",
       "      <td>14.285714</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>True</td>\n",
       "      <td>3020945.0</td>\n",
       "      <td>16.666667</td>\n",
       "      <td>2.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>True</td>\n",
       "      <td>15.476190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1</td>\n",
       "      <td>3020946</td>\n",
       "      <td>3020946.0</td>\n",
       "      <td>58.333333</td>\n",
       "      <td>7.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>True</td>\n",
       "      <td>3020946.0</td>\n",
       "      <td>66.666667</td>\n",
       "      <td>10.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>True</td>\n",
       "      <td>62.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1</td>\n",
       "      <td>3020971</td>\n",
       "      <td>3020971.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>3020971.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1</td>\n",
       "      <td>3020972</td>\n",
       "      <td>3020972.0</td>\n",
       "      <td>75.000000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>3020972.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>87.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1</td>\n",
       "      <td>3020987</td>\n",
       "      <td>3020987.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>3020987.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1</td>\n",
       "      <td>3020988</td>\n",
       "      <td>3020988.0</td>\n",
       "      <td>75.000000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "      <td>3020988.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>87.500000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chromosome    s_loc  e_loc_df1  methyl_rate_df1  s_depth_df1  e_depth_df1  \\\n",
       "0           1  3007431  3007431.0         0.000000          0.0          1.0   \n",
       "1           1  3014611  3014611.0       100.000000          1.0          0.0   \n",
       "2           1  3014612  3014612.0         7.692308          2.0         24.0   \n",
       "3           1  3014928  3014928.0       100.000000          1.0          0.0   \n",
       "4           1  3014929  3014929.0        76.000000         19.0          6.0   \n",
       "5           1  3014974  3014974.0         0.000000          0.0          1.0   \n",
       "6           1  3014975  3014975.0        92.592593         25.0          2.0   \n",
       "7           1  3020794  3020794.0       100.000000          1.0          0.0   \n",
       "8           1  3020814  3020814.0       100.000000          1.0          0.0   \n",
       "9           1  3020842  3020842.0         0.000000          0.0          1.0   \n",
       "10          1  3020877  3020877.0        80.000000          4.0          1.0   \n",
       "11          1  3020878  3020878.0         0.000000          0.0          8.0   \n",
       "12          1  3020891  3020891.0        40.000000          2.0          3.0   \n",
       "13          1  3020892  3020892.0        75.000000          6.0          2.0   \n",
       "14          1  3020945  3020945.0        14.285714          1.0          6.0   \n",
       "15          1  3020946  3020946.0        58.333333          7.0          5.0   \n",
       "16          1  3020971  3020971.0       100.000000          2.0          0.0   \n",
       "17          1  3020972  3020972.0        75.000000          3.0          1.0   \n",
       "18          1  3020987  3020987.0       100.000000          2.0          0.0   \n",
       "19          1  3020988  3020988.0        75.000000          3.0          1.0   \n",
       "\n",
       "   CG site_df1  e_loc_df2  methyl_rate_df2  s_depth_df2  e_depth_df2  \\\n",
       "0        False  3007431.0         0.000000          0.0          1.0   \n",
       "1         True  3014611.0       100.000000          2.0          0.0   \n",
       "2         True  3014612.0         0.000000          0.0         11.0   \n",
       "3         True  3014928.0         0.000000          0.0          1.0   \n",
       "4         True  3014929.0        90.000000          9.0          1.0   \n",
       "5         True  3014974.0         0.000000          0.0          2.0   \n",
       "6         True  3014975.0        90.909091         10.0          1.0   \n",
       "7        False        NaN              NaN          NaN          NaN   \n",
       "8        False        NaN              NaN          NaN          NaN   \n",
       "9        False        NaN              NaN          NaN          NaN   \n",
       "10        True  3020877.0       100.000000          7.0          0.0   \n",
       "11        True  3020878.0         0.000000          0.0         11.0   \n",
       "12        True  3020891.0       100.000000          9.0          0.0   \n",
       "13        True  3020892.0       100.000000         11.0          0.0   \n",
       "14        True  3020945.0        16.666667          2.0         10.0   \n",
       "15        True  3020946.0        66.666667         10.0          5.0   \n",
       "16        True  3020971.0       100.000000          3.0          0.0   \n",
       "17        True  3020972.0       100.000000          4.0          0.0   \n",
       "18        True  3020987.0       100.000000          3.0          0.0   \n",
       "19        True  3020988.0       100.000000          4.0          0.0   \n",
       "\n",
       "   CG site_df2  avg_methyl_rate  \n",
       "0        False         0.000000  \n",
       "1         True       100.000000  \n",
       "2         True         3.846154  \n",
       "3         True        50.000000  \n",
       "4         True        83.000000  \n",
       "5         True         0.000000  \n",
       "6         True        91.750842  \n",
       "7          NaN       100.000000  \n",
       "8          NaN       100.000000  \n",
       "9          NaN         0.000000  \n",
       "10        True        90.000000  \n",
       "11        True         0.000000  \n",
       "12        True        70.000000  \n",
       "13        True        87.500000  \n",
       "14        True        15.476190  \n",
       "15        True        62.500000  \n",
       "16        True       100.000000  \n",
       "17        True        87.500000  \n",
       "18        True       100.000000  \n",
       "19        True        87.500000  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "data/GSE93957_RAW/GSM2465662_M0420527_14wk_Heart.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465658_M0420525_14wk_Heart.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465674_M0520522_14wk_Heart.cov.txt.gz\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "merge() missing 1 required positional argument: 'right'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m/home/hunter/Programming/Bioinformatics/Projects/Methylation_1/Bioinformatic_Lab-1/test.ipynb Cell 12\u001b[0m line \u001b[0;36m2\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/hunter/Programming/Bioinformatics/Projects/Methylation_1/Bioinformatic_Lab-1/test.ipynb#X14sZmlsZQ%3D%3D?line=19'>20</a>\u001b[0m     \u001b[39melse\u001b[39;00m:\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/hunter/Programming/Bioinformatics/Projects/Methylation_1/Bioinformatic_Lab-1/test.ipynb#X14sZmlsZQ%3D%3D?line=20'>21</a>\u001b[0m         \u001b[39mraise\u001b[39;00m \u001b[39mException\u001b[39;00m(\u001b[39m\"\u001b[39m\u001b[39mno dataframes found\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[0;32m---> <a href='vscode-notebook-cell:/home/hunter/Programming/Bioinformatics/Projects/Methylation_1/Bioinformatic_Lab-1/test.ipynb#X14sZmlsZQ%3D%3D?line=24'>25</a>\u001b[0m merged_df \u001b[39m=\u001b[39m grab_tissue(\u001b[39m\"\u001b[39;49m\u001b[39mHeart\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39m\"\u001b[39;49m\u001b[39m14wk\u001b[39;49m\u001b[39m\"\u001b[39;49m)\u001b[39m.\u001b[39mhead(\u001b[39m10\u001b[39m)\n",
      "\u001b[1;32m/home/hunter/Programming/Bioinformatics/Projects/Methylation_1/Bioinformatic_Lab-1/test.ipynb Cell 12\u001b[0m line \u001b[0;36m1\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/hunter/Programming/Bioinformatics/Projects/Methylation_1/Bioinformatic_Lab-1/test.ipynb#X14sZmlsZQ%3D%3D?line=14'>15</a>\u001b[0m     suffixes\u001b[39m.\u001b[39mappend(\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m_df\u001b[39m\u001b[39m{\u001b[39;00mi\u001b[39m}\u001b[39;00m\u001b[39m\"\u001b[39m)\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/hunter/Programming/Bioinformatics/Projects/Methylation_1/Bioinformatic_Lab-1/test.ipynb#X14sZmlsZQ%3D%3D?line=16'>17</a>\u001b[0m \u001b[39mif\u001b[39;00m dataframes_to_merge:\n\u001b[0;32m---> <a href='vscode-notebook-cell:/home/hunter/Programming/Bioinformatics/Projects/Methylation_1/Bioinformatic_Lab-1/test.ipynb#X14sZmlsZQ%3D%3D?line=17'>18</a>\u001b[0m     merged_df \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39;49mmerge(dataframes_to_merge, on\u001b[39m=\u001b[39;49m[\u001b[39m'\u001b[39;49m\u001b[39mchromosome\u001b[39;49m\u001b[39m'\u001b[39;49m, \u001b[39m'\u001b[39;49m\u001b[39ms_loc\u001b[39;49m\u001b[39m'\u001b[39;49m], how\u001b[39m=\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39mouter\u001b[39;49m\u001b[39m\"\u001b[39;49m, suffixes\u001b[39m=\u001b[39;49msuffixes, sort\u001b[39m=\u001b[39;49m\u001b[39mTrue\u001b[39;49;00m)\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/hunter/Programming/Bioinformatics/Projects/Methylation_1/Bioinformatic_Lab-1/test.ipynb#X14sZmlsZQ%3D%3D?line=18'>19</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m merged_df\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/hunter/Programming/Bioinformatics/Projects/Methylation_1/Bioinformatic_Lab-1/test.ipynb#X14sZmlsZQ%3D%3D?line=19'>20</a>\u001b[0m \u001b[39melse\u001b[39;00m:\n",
      "\u001b[0;31mTypeError\u001b[0m: merge() missing 1 required positional argument: 'right'"
     ]
    }
   ],
   "source": [
    "def grab_tissue(tissue_type: str, age: str):\n",
    "    merged_df =  pd.DataFrame(columns=['chromosome', 's_loc', 'e_loc', 'methyl_rate', 's_depth', 'e_depth'])\n",
    "    # Initialize an empty list to store individual DataFrames\n",
    "    dataframes_to_merge = []\n",
    "\n",
    "\n",
    "    matching_tissue = os.listdir(directory)\n",
    "    for tissue in matching_tissue:\n",
    "        if age in tissue and tissue_type in tissue:\n",
    "            df = build_df(tissue)\n",
    "            dataframes_to_merge.append(df)\n",
    "    \n",
    "    suffixes = []\n",
    "    for i in range(len(dataframes_to_merge)):\n",
    "        suffixes.append(f\"_df{i}\")\n",
    "\n",
    "    if dataframes_to_merge:\n",
    "        merged_df = pd.merge(dataframes_to_merge, on=['chromosome', 's_loc'], how=\"outer\", suffixes=suffixes, sort=True)\n",
    "        return merged_df\n",
    "    else:\n",
    "        raise Exception(\"no dataframes found\")\n",
    "\n",
    "            \n",
    "\n",
    "merged_df = grab_tissue(\"Heart\", \"14wk\").head(10)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "data/GSE93957_RAW/GSM2465662_M0420527_14wk_Heart.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465658_M0420525_14wk_Heart.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465674_M0520522_14wk_Heart.cov.txt.gz\n"
     ]
    }
   ],
   "source": [
    "def grab_tissue2(tissue_type: str, age: str):\n",
    "    merged_df =  pd.DataFrame(columns=['chromosome', 's_loc', 'e_loc', 'methyl_rate', 's_depth', 'e_depth'])\n",
    "    # Initialize an empty list to store individual DataFrames\n",
    "    dataframes_to_merge = []\n",
    "    matching_tissue = os.listdir(directory)\n",
    "\n",
    "    for tissue in matching_tissue:\n",
    "        if age in tissue and tissue_type in tissue:\n",
    "            df = build_df(tissue)\n",
    "            dataframes_to_merge.append(df)\n",
    "\n",
    "    # Check if there are any DataFrames to merge\n",
    "    if dataframes_to_merge:\n",
    "        # Initialize the merged DataFrame with the first DataFrame\n",
    "        merged_df = dataframes_to_merge[0]\n",
    "\n",
    "        # Iterate through the rest of the DataFrames and merge them with a unique suffix\n",
    "        for i, df in enumerate(dataframes_to_merge[1:], start=2):\n",
    "            print(\"test\")\n",
    "            suffix = f\"_df{i}\"  # Create a unique suffix based on the DataFrame index\n",
    "            merged_df = merged_df.merge(df, on=['chromosome', 's_loc'], suffixes=('', suffix))\n",
    "    else:\n",
    "        raise Exception(\"no dataframes found\")\n",
    "    return merged_df\n",
    "\n",
    "grab_tissue2(\"Heart\", \"14wk\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'merged_df' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m/home/hunter/Programming/Bioinformatics/Projects/Methylation_1/Bioinformatic_Lab-1/test.ipynb Cell 14\u001b[0m line \u001b[0;36m1\n\u001b[0;32m----> <a href='vscode-notebook-cell:/home/hunter/Programming/Bioinformatics/Projects/Methylation_1/Bioinformatic_Lab-1/test.ipynb#X20sZmlsZQ%3D%3D?line=0'>1</a>\u001b[0m methyl_rate_columns \u001b[39m=\u001b[39m [col \u001b[39mfor\u001b[39;00m col \u001b[39min\u001b[39;00m merged_df\u001b[39m.\u001b[39mcolumns \u001b[39mif\u001b[39;00m col\u001b[39m.\u001b[39mstartswith(\u001b[39m'\u001b[39m\u001b[39mmethyl_rate_\u001b[39m\u001b[39m'\u001b[39m)]\n\u001b[1;32m      <a href='vscode-notebook-cell:/home/hunter/Programming/Bioinformatics/Projects/Methylation_1/Bioinformatic_Lab-1/test.ipynb#X20sZmlsZQ%3D%3D?line=1'>2</a>\u001b[0m merged_df[\u001b[39m'\u001b[39m\u001b[39mavg_methyl_rate\u001b[39m\u001b[39m'\u001b[39m] \u001b[39m=\u001b[39m merged_df[methyl_rate_columns]\u001b[39m.\u001b[39mmean(axis\u001b[39m=\u001b[39m\u001b[39m1\u001b[39m)\n\u001b[1;32m      <a href='vscode-notebook-cell:/home/hunter/Programming/Bioinformatics/Projects/Methylation_1/Bioinformatic_Lab-1/test.ipynb#X20sZmlsZQ%3D%3D?line=2'>3</a>\u001b[0m merged_df\u001b[39m.\u001b[39mhead()\n",
      "\u001b[0;31mNameError\u001b[0m: name 'merged_df' is not defined"
     ]
    }
   ],
   "source": [
    "methyl_rate_columns = [col for col in merged_df.columns if col.startswith('methyl_rate_')]\n",
    "merged_df['avg_methyl_rate'] = merged_df[methyl_rate_columns].mean(axis=1)\n",
    "merged_df.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "vmeth",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
