{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### WIP Methylation Statistical Analysis\n",
    "\n",
    "This code analyzes a multi-tissue DNA methylation dataset from mice, sourced from ['Multi-tissue DNA methylation age predictor in mouse'](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-017-1203-5). It calculates average methylation rate, depth, and identifies tissue-specific and age-related differentiations in the samples. By employing  statistical techniques, this code will provide valuable insights into DNA methylation patterns and their association with aging processes. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "import os\n",
    "import pandas as pd\n",
    "import sys\n",
    "from typing import Tuple, List"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "directory = 'data/GSE93957_RAW/'"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load Data & Build Dataframe\n",
    "\n",
    "The build_df function generates a pandas DataFrame from a specified sample name. It initializes an empty DataFrame and defines column names. It iterates through files in the directory, checking for gzip format and matching sample name. Data is read using read_csv, with tab as separator, and stored in a temporary DataFrame. If the main DataFrame is empty, the data is directly assigned. The function returns the compiled DataFrame from matching files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "sample_list = [\"Lung\", \"Heart\", \"Liver\", \"Cortex\"]\n",
    "\n",
    "def build_df(filename) -> Tuple[pd.DataFrame, str]:\n",
    "    # Create empty DataFrame to be filled\n",
    "    df = pd.DataFrame()\n",
    "    columns_names = ['chromosome', 's_loc', 'e_loc', 'methyl_rate', 's_depth', 'e_depth']\n",
    "    \n",
    "    # # Choose preset file to perform statistical analysis on\n",
    "    # if test:\n",
    "    #     filename = \"GSM2465667_M04NB_1wk_Liver.cov.txt.gz\"\n",
    "    #     columns_names = ['chromosome', 's_loc', 'e_loc', 'methyl_rate', 's_depth', 'e_depth']\n",
    "    #     with gzip.open('data/GSE93957_RAW/GSM2465667_M04NB_1wk_Liver.cov.txt.gz', 'rt') as file: \n",
    "    #         df = pd.read_csv('data/GSE93957_RAW/GSM2465667_M04NB_1wk_Liver.cov.txt.gz', sep=\"\\t\", header=None, names=columns_names, low_memory=False)\n",
    "    #     return df, filename\n",
    "\n",
    "\n",
    "    file_path = os.path.join(directory, filename)\n",
    "    # Open the gzipped file in text mode\n",
    "    with gzip.open(file_path, 'rt') as file:  \n",
    "        print(file_path)\n",
    "        df = pd.read_csv(file_path, sep=\"\\t\", header=None, names=columns_names, low_memory=False)\n",
    "      \n",
    "    return df, filename"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Build Final Result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_result(test = False):\n",
    "    column_names = ['sites', 'ave depth', 'ave methylation', 'ave methylation > 2 depth', 'ave methylation > 5 depth']\n",
    "    result = pd.DataFrame(columns=column_names)\n",
    "    \n",
    "    \n",
    "    if not test:\n",
    "        for filename in os.listdir(directory):\n",
    "            df, name = build_df(filename)\n",
    "            length = len(df)\n",
    "            depth = (df['s_depth'] + df['e_depth']).mean()\n",
    "            mean = df['methyl_rate'].mean()\n",
    "            \n",
    "            # Get the average methylation for rows with depth >= 2\n",
    "            df = df.drop(df[df['e_depth'] + df['s_depth'] < 2].index)\n",
    "            df = df.drop(df[df['e_depth'] + df['s_depth'] > 100].index)\n",
    "            two_depth_mean = df['methyl_rate'].mean()\n",
    "\n",
    "            df = df.drop(df[df['e_depth'] + df['s_depth'] < 5].index)\n",
    "            five_depth_mean = df['methyl_rate'].mean()\n",
    "            result.loc[len(result)] = [length, depth, mean, two_depth_mean, five_depth_mean]\n",
    "    else: \n",
    "        build_df(filename=\"GSM2465667_M04NB_1wk_Liver.cov.txt.gz\")\n",
    "        length = len(df)\n",
    "        depth = (df['s_depth'] + df['e_depth']).mean()\n",
    "        mean = df['methyl_rate'].mean()\n",
    "        \n",
    "        # Get the average methylation for rows with depth >= 2\n",
    "        df = df.drop(df[df['e_depth'] + df['s_depth'] < 2].index)\n",
    "        df = df.drop(df[df['e_depth'] + df['s_depth'] > 100].index)\n",
    "        two_depth_mean = df['methyl_rate'].mean()\n",
    "\n",
    "        df = df.drop(df[df['e_depth'] + df['s_depth'] < 5].index)\n",
    "        five_depth_mean = df['methyl_rate'].mean()\n",
    "        result.loc[len(result)] = [length, depth, mean, two_depth_mean, five_depth_mean]\n",
    "        \n",
    "\n",
    "    return result, name\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Display menu for user\n",
    "\n",
    "In this example we choose '5': all samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "data/GSE93957_RAW/GSM2465633_M00018724_27wk_Liver.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465619_M00018362_41wk_Cortex.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465667_M04NB_1wk_Liver.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465627_M00018381_41wk_Cortex.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465631_M00018724_27wk_Cortex.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465653_M02NB_1wk_Lung.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465636_M00018752_27wk_Heart.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465632_M00018724_27wk_Heart.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465637_M00018752_27wk_Liver.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465625_M00018363_41wk_Liver.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465668_M04NB_1wk_Lung.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465662_M0420527_14wk_Heart.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465665_M04NB_1wk_Cortex.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465642_M00018754_27wk_Lung.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465676_M0520522_14wk_Lung.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465621_M00018362_41wk_Liver.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465675_M0520522_14wk_Liver.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465648_M01NB_1wk_Heart.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465640_M00018754_27wk_Heart.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465646_M0112371_27wk_Lung.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465664_M0420527_14wk_Lung.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465629_M00018381_41wk_Liver.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465656_M03NB_1wk_Lung.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465659_M0420525_14wk_Liver.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465654_M03NB_1wk_Cortex.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465650_M01NB_1wk_Lung.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465663_M0420527_14wk_Liver.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465657_M0420525_14wk_Cortex.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465670_M0512375_27wk_Heart.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465626_M00018363_41wk_Lung.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465623_M00018363_41wk_Cortex.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465620_M00018362_41wk_Heart.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465658_M0420525_14wk_Heart.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465672_M0512375_27wk_Lung.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465644_M0112371_27wk_Heart.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465669_M0512375_27wk_Cortex.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465674_M0520522_14wk_Heart.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465624_M00018363_41wk_Heart.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465634_M00018724_27wk_Lung.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465652_M02NB_1wk_Liver.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465645_M0112371_27wk_Liver.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465630_M00018381_41wk_Lung.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465639_M00018754_27wk_Cortex.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465635_M00018752_27wk_Cortex.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465616_M00018359_41wk_Heart.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465638_M00018752_27wk_Lung.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465673_M0520522_14wk_Cortex.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465651_M02NB_1wk_Cortex.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465641_M00018754_27wk_Liver.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465643_M0112371_27wk_Cortex.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465661_M0420527_14wk_Cortex.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465622_M00018362_41wk_Lung.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465615_M00018359_41wk_Cortex.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465617_M00018359_41wk_Liver.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465660_M0420525_14wk_Lung.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465618_M00018359_41wk_Lung.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465671_M0512375_27wk_Liver.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465628_M00018381_41wk_Heart.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465647_M01NB_1wk_Cortex.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465649_M01NB_1wk_Liver.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465666_M04NB_1wk_Heart.cov.txt.gz\n",
      "data/GSE93957_RAW/GSM2465655_M03NB_1wk_Heart.cov.txt.gz\n"
     ]
    }
   ],
   "source": [
    "# Perform statistical analysis and build a 'result DataFrame' that will be saved\n",
    "df, row_name = build_result()\n",
    "df['Row'] = row_name\n",
    "column_order = ['Row', 'sites', 'ave depth', 'ave methylation', 'ave methylation > 2 depth', 'ave methylation > 5 depth']\n",
    "df = df[column_order]\n",
    "df.to_csv('result.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Row</th>\n",
       "      <th>sites</th>\n",
       "      <th>ave depth</th>\n",
       "      <th>ave methylation</th>\n",
       "      <th>ave methylation &gt; 2 depth</th>\n",
       "      <th>ave methylation &gt; 5 depth</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GSM2465655_M03NB_1wk_Heart.cov.txt.gz</td>\n",
       "      <td>4748562.0</td>\n",
       "      <td>11.236037</td>\n",
       "      <td>42.379657</td>\n",
       "      <td>36.123860</td>\n",
       "      <td>36.826557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GSM2465655_M03NB_1wk_Heart.cov.txt.gz</td>\n",
       "      <td>4121662.0</td>\n",
       "      <td>14.977096</td>\n",
       "      <td>38.715772</td>\n",
       "      <td>37.048571</td>\n",
       "      <td>38.045658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GSM2465655_M03NB_1wk_Heart.cov.txt.gz</td>\n",
       "      <td>4147692.0</td>\n",
       "      <td>13.342087</td>\n",
       "      <td>36.218352</td>\n",
       "      <td>34.165614</td>\n",
       "      <td>35.222294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GSM2465655_M03NB_1wk_Heart.cov.txt.gz</td>\n",
       "      <td>4165468.0</td>\n",
       "      <td>15.123178</td>\n",
       "      <td>37.681559</td>\n",
       "      <td>36.102748</td>\n",
       "      <td>37.034388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GSM2465655_M03NB_1wk_Heart.cov.txt.gz</td>\n",
       "      <td>4086565.0</td>\n",
       "      <td>15.372909</td>\n",
       "      <td>38.633424</td>\n",
       "      <td>37.273794</td>\n",
       "      <td>39.312597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>GSM2465655_M03NB_1wk_Heart.cov.txt.gz</td>\n",
       "      <td>3917926.0</td>\n",
       "      <td>12.731100</td>\n",
       "      <td>35.657432</td>\n",
       "      <td>34.545901</td>\n",
       "      <td>35.690143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>GSM2465655_M03NB_1wk_Heart.cov.txt.gz</td>\n",
       "      <td>3939791.0</td>\n",
       "      <td>9.395831</td>\n",
       "      <td>36.255070</td>\n",
       "      <td>34.480659</td>\n",
       "      <td>35.639115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>GSM2465655_M03NB_1wk_Heart.cov.txt.gz</td>\n",
       "      <td>3669961.0</td>\n",
       "      <td>7.935134</td>\n",
       "      <td>34.789121</td>\n",
       "      <td>34.387766</td>\n",
       "      <td>36.722110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>GSM2465655_M03NB_1wk_Heart.cov.txt.gz</td>\n",
       "      <td>4257344.0</td>\n",
       "      <td>19.341303</td>\n",
       "      <td>34.964655</td>\n",
       "      <td>32.868154</td>\n",
       "      <td>33.876040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>GSM2465655_M03NB_1wk_Heart.cov.txt.gz</td>\n",
       "      <td>3864703.0</td>\n",
       "      <td>14.244776</td>\n",
       "      <td>34.843568</td>\n",
       "      <td>34.219407</td>\n",
       "      <td>36.359592</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      Row      sites  ave depth  \\\n",
       "0   GSM2465655_M03NB_1wk_Heart.cov.txt.gz  4748562.0  11.236037   \n",
       "1   GSM2465655_M03NB_1wk_Heart.cov.txt.gz  4121662.0  14.977096   \n",
       "2   GSM2465655_M03NB_1wk_Heart.cov.txt.gz  4147692.0  13.342087   \n",
       "3   GSM2465655_M03NB_1wk_Heart.cov.txt.gz  4165468.0  15.123178   \n",
       "4   GSM2465655_M03NB_1wk_Heart.cov.txt.gz  4086565.0  15.372909   \n",
       "..                                    ...        ...        ...   \n",
       "57  GSM2465655_M03NB_1wk_Heart.cov.txt.gz  3917926.0  12.731100   \n",
       "58  GSM2465655_M03NB_1wk_Heart.cov.txt.gz  3939791.0   9.395831   \n",
       "59  GSM2465655_M03NB_1wk_Heart.cov.txt.gz  3669961.0   7.935134   \n",
       "60  GSM2465655_M03NB_1wk_Heart.cov.txt.gz  4257344.0  19.341303   \n",
       "61  GSM2465655_M03NB_1wk_Heart.cov.txt.gz  3864703.0  14.244776   \n",
       "\n",
       "    ave methylation  ave methylation > 2 depth  ave methylation > 5 depth  \n",
       "0         42.379657                  36.123860                  36.826557  \n",
       "1         38.715772                  37.048571                  38.045658  \n",
       "2         36.218352                  34.165614                  35.222294  \n",
       "3         37.681559                  36.102748                  37.034388  \n",
       "4         38.633424                  37.273794                  39.312597  \n",
       "..              ...                        ...                        ...  \n",
       "57        35.657432                  34.545901                  35.690143  \n",
       "58        36.255070                  34.480659                  35.639115  \n",
       "59        34.789121                  34.387766                  36.722110  \n",
       "60        34.964655                  32.868154                  33.876040  \n",
       "61        34.843568                  34.219407                  36.359592  \n",
       "\n",
       "[62 rows x 6 columns]"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Display Results Table\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "36.96097574055515"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['ave methylation > 5 depth'].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "vmeth",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
